# 3量子ビットゲート
前回までは1量子ビットと2量子ビットの演算を行なってきました。トフォリゲートは3量子ビットを利用する演算です。トフォリゲートはトマソ・トフォリによって作られ、二つのコントロールビットと一つのターゲットビットを持ちます。両方のコントロールビットが1の時にのみ、ターゲットビットにXゲートが適用されます。早速みてみましょう。

まずはインストールをします。

In [ ]:
!pip install blueqat

## 回路の作成
トフォリゲートは利用するのは簡単です。.ccxもしくは.toffoliを使います。
まずは単純に回路を書いてみましょう。ここでは*をコントロールビット、Xをターゲットビットとしています。

<img src="./img/005/005_0.png" width="15%">

トフォリゲートは、





In [3]:
from blueqat import Circuit
Circuit().ccx[0,1,2].m[:].run(shots=1)

Counter({'000': 1})

もしくは、

In [4]:
Circuit().toffoli[0,1,2].m[:].run(shots=1)

Counter({'000': 1})

のように書けます。どちらの書き方をつかっても大丈夫です。上記はコントロールビットが両方とも0なので、ターゲットビットには変化がありません。

##コントロールビットを変化させてみる
コントロールビットの片方を1にした時と、両方を1にした時とでどうなるかみてみましょう。最初にコントロールビットの両方を1にしてみます。1にするには反転ゲートのXゲートを使います。

In [9]:
Circuit().x[0,1].ccx[0,1,2].m[:].run(shots=1)

Counter({'111': 1})

どうなったでしょうか。0番目と1番目の量子ビットを1にしたことにより、2番目の量子ビットが0から1に変化しました。このようにコントロールビットを両方とも1にするとターゲットビットが反転します。

コントロールビットの片方の量子ビットだけを1にしたときはどうでしょうか。

In [10]:
Circuit().x[0].ccx[0,1,2].m[:].run(shots=1)

Counter({'100': 1})

こちらもターゲットビットは変化なしです。

In [11]:
Circuit().x[1].ccx[0,1,2].m[:].run(shots=1)

Counter({'010': 1})

こちらもコントロールビットの片方だけが1なのでターゲットビットは反応なしです。

##トフォリゲートの真理値表と行列表現
トフォリゲートの入出力は真理値表というもので確認ができます。

入力 | 出力
--- | ---
000 | 000
001 | 001
010 | 010
011 | 011
100 | 100
101 | 101
110 | 111
111 | 110



これをみると、110と111の時だけターゲットビットが反転しています。ゲートの行列表現も見てみましょう。

$\begin{bmatrix}
1&0&0&0&0&0&0&0\\
0&1&0&0&0&0&0&0\\
0&0&1&0&0&0&0&0\\
0&0&0&1&0&0&0&0\\
0&0&0&0&1&0&0&0\\
0&0&0&0&0&1&0&0\\
0&0&0&0&0&0&0&1\\
0&0&0&0&0&0&1&0\\
\end{bmatrix}$


最後のゲート演算でXゲートが見れます。真理値表とも合っているのが確認できます。

##トフォリゲートの中身
トフォリゲートは実は中身は1量子ビットゲートと2量子ビットゲートで構成されています。回路はHとCXとTで構成されます。ちょっと長いですががんばってかいてみましょう。

In [12]:
from blueqat import Circuit
import numpy as np

Circuit().x[1:].h[0].cnot[1,0].rz(-np.pi/4)[0].cnot[2,0].rz(np.pi/4)[0].cnot[1,0].rz(-np.pi/4)[0].cnot[2,0].rz(np.pi/4)[:1].h[0].cnot[1,0].cnot[0,1].cnot[1,0].cnot[2,0].rz(-np.pi/4)[0].rz(np.pi/4)[2].cnot[2,0].m[:].run(shots=1)

Counter({'111': 1})

回路は以下のようになります。

<img src="./img/005/005_1.png" width="85%">